### PR TITLE
adding API v20170930 version for upgrade operation

### PR DESCRIPTION
**Pull request by dmitsh**
adding new classes and API for upgrade operation (unversioned, vlabs, v20170930)

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/1366)

### CODE REVIEW

File: pkg/acsengine/orchestrators.go [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/Azure/acs-engine/pkg/api"
+	"github.com/Azure/acs-engine/pkg/api/common"
+)
+
+// OrchestratorsInfo contains list of release info for supported orchestrators
+type OrchestratorsInfo struct {
+	Orchestrators []*api.OrchestratorInfo `json:"orchestrators"`
```
Comment (weinong) [inline pkg/acsengine/orchestrators.go]: should this be named as orchestratorInfos

File: pkg/acsengine/orchestrators.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Azure/acs-engine/pkg/api"
+	"github.com/Azure/acs-engine/pkg/api/common"
+)
+
+// OrchestratorsInfo contains list of release info for supported orchestrators
+type OrchestratorsInfo struct {
```
Comment (weinong) [inline pkg/acsengine/orchestrators.go]: OrchestratorInfos ?

File: pkg/acsengine/orchestrators.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+			Orchestrator: api.Swarm,
+			VersionInfo: api.VersionInfo{
+				Version: SwarmVersion,
+			},
+			DockerComposeVersion: SwarmDockerComposeVersion,
+		},
+	}, nil
+}
+
+func dockerceInfo(release string) ([]*api.OrchestratorInfo, error) {
```
Comment (weinong) [inline pkg/acsengine/orchestrators.go]: does it not need `Default`?

Comment (weinong) [reply pkg/acsengine/orchestrators.go]: and swarmInfo above

File: pkg/acsengine/orchestrators.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+	return []*api.OrchestratorInfo{
+		{
+			Orchestrator: api.SwarmMode,
```
Comment (weinong) [inline pkg/acsengine/orchestrators.go]: where does the conversion back to DocerCE happen?

File: pkg/api/types.go [hunk 5]
```diff
@@ -139,6 +139,21 @@ type OrchestratorProfile struct {
...
+type VersionInfo struct {
+	Release string `json:"release,omitempty"`
+	Version string `json:"version"`
+}
+
+// OrchestratorInfo contains orchestrator version info
+type OrchestratorInfo struct {
+	Orchestrator string `json:"orchestrator"`
+	VersionInfo
+	DockerComposeVersion string         `json:"docker-compose-version,omitempty"`
```
Comment (weinong) [inline pkg/api/types.go]: I thought the convention is to use camel casing.

File: pkg/acsengine/orchestrators.go (hunk 3 above)
Comment (dmitsh) [reply pkg/acsengine/orchestrators.go]: no, we support only one version.

Comment (dmitsh) [reply pkg/acsengine/orchestrators.go]: unless that should make it default ... :)

File: pkg/acsengine/orchestrators.go (hunk 1 above)
Comment (weinong) [reply pkg/acsengine/orchestrators.go]:
```
type OrchestratorInfosResult struct {
   OrchestratorInfos []*api.OrchestratorInfo `json:"orchestratorInfos"`
}
```

Comment (dmitsh) [reply pkg/acsengine/orchestrators.go]: changed

File: cmd/orchestrators.go [hunk 6]
```diff
@@ -47,91 +39,15 @@ func newOrchestratorsCmd() *cobra.Command {
...
-		}
-	default:
-		return fmt.Errorf("Unsupported orchestrator '%s'", oc.orchestrator)
-	}
-	return nil
-}
-
 func (oc *orchestratorsCmd) run(cmd *cobra.Command, args []string) error {
-	if err := oc.validate(); err != nil {
+	orchs, err := acsengine.NewOrchestratorInfos(oc.orchestrator, oc.release)
```
Comment (amanohar) [inline cmd/orchestrators.go]: I don't think 's' is needed in the name here.

File: pkg/api/vlabs/types.go [hunk 7]
```diff
@@ -143,6 +143,21 @@ type OrchestratorProfile struct {
 	KubernetesConfig    *KubernetesConfig `json:"kubernetesConfig,omitempty"`
 }
 
+// VersionInfo contains rersion and release numbers
```
Comment (amanohar) [inline pkg/api/vlabs/types.go]: This file is not the right place for these structs. Can be in upgradetypesandvalidate or in a new file

File: pkg/api/v20170930/types.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package v20170930
+
+// VersionInfo contains rersion and release numbers
+type VersionInfo struct {
+	Release string `json:"release,omitempty"`
```
Comment (amanohar) [inline pkg/api/v20170930/types.go]: Why not be consistent and use?:
OrchestratorRelease string            `json:"orchestratorRelease"`
OrchestratorVersion string            `json:"orchestratorVersion" validate:"len=0"`

File: pkg/api/v20170930/types.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+	Version string `json:"version"`
+}
+
+// OrchestratorInfo contains orchestrator version info
+type OrchestratorInfo struct {
+	Orchestrator string `json:"orchestrator"`
+	VersionInfo
+	DockerComposeVersion string         `json:"docker-compose-version,omitempty"`
+	Default              bool           `json:"default,omitempty"`
```
Comment (amanohar) [inline pkg/api/v20170930/types.go]: Please add comments describing what Default means in the context here..

File: pkg/api/v20170930/types.go [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+type VersionInfo struct {
+	Release string `json:"release,omitempty"`
+	Version string `json:"version"`
+}
+
+// OrchestratorInfo contains orchestrator version info
+type OrchestratorInfo struct {
+	Orchestrator string `json:"orchestrator"`
+	VersionInfo
+	DockerComposeVersion string         `json:"docker-compose-version,omitempty"`
```
Comment (amanohar) [inline pkg/api/v20170930/types.go]: Do we really need to expose this as a field in the RP? IMO no just like we don't expose heapster or etcd version.

File: pkg/api/v20170930/types.go [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+// VersionInfo contains rersion and release numbers
+type VersionInfo struct {
+	Release string `json:"release,omitempty"`
+	Version string `json:"version"`
+}
+
+// OrchestratorInfo contains orchestrator version info
+type OrchestratorInfo struct {
+	Orchestrator string `json:"orchestrator"`
+	VersionInfo
```
Comment (amanohar) [inline pkg/api/v20170930/types.go]: If Orchestrator is a separate field does VersionInfo need to a struct? In general the organization of this struct is confusing. Why result orchestratorProfile here?

File: pkg/acsengine/orchestrators.go [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+	if cs == nil || cs.Properties == nil || cs.Properties.OrchestratorProfile == nil {
+		return nil, fmt.Errorf("Incomplete ContainerService")
+	}
+	if len(cs.Properties.OrchestratorProfile.OrchestratorRelease) == 0 {
+		return nil, fmt.Errorf("Missing Orchestrator Release")
+	}
+	if cs.Properties.OrchestratorProfile.OrchestratorType != api.Kubernetes {
+		return nil, fmt.Errorf("Upgrade operation is not supported for '%s'", cs.Properties.OrchestratorProfile.OrchestratorType)
+	}
+	arr, err := kubernetesInfo(&api.VersionInfo{
```
Comment (amanohar) [inline pkg/acsengine/orchestrators.go]: As noted in general it's not clear why another struct is needed to denote the same info as orchestratorProfile. This looks like a duplication that can be avoided.

File: pkg/acsengine/orchestrators.go [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+	if err := csVer.parse(csInfo.Version); err != nil {
+		return ret, err
+	}
+	switch csInfo.Release {
+	case common.KubernetesRelease1Dot5:
+		// add next release
+		ret = append(ret, &api.VersionInfo{Release: common.KubernetesRelease1Dot6, Version: common.KubeReleaseToVersion[common.KubernetesRelease1Dot6]})
+	case common.KubernetesRelease1Dot6:
+		// check for patch upgrade
+		patchVersion := common.KubeReleaseToVersion[common.KubernetesRelease1Dot6]
```
Comment (amanohar) [inline pkg/acsengine/orchestrators.go]: Looks like a lot of code repetition within each version. Can be extracted and reused?

**Comment by sauryadas** [comment]
@dmitsh how can we try this out ?

**Pull request merged**